runbotics/runbotics-scheduler/src/queue/processor/scheduler.processor.ts

```diff
@@ -65,28 +71,108 @@ export class SchedulerProcessor {
+
+        const active = await this.queueService.getActiveJobs();
+        const waiting = await this.queueService.getWaitingJobs();
+        const jobs = [...active, ...waiting];
+
+        const jobIndex = jobs.findIndex((job) => job.id === jobId);
+        const job = jobs[jobIndex];
+        if (!job) return;
+
+        this.uiGateway.server.emit(WsMessage.ADD_WAITING_SCHEDULE, job);
+
+        const queuePosition = jobIndex + 1;
+        const processId = job?.data?.process?.id;
+        const orchestratorProcessInstanceId = job?.data?.orchestratorProcessInstanceId;
+        this.uiGateway.emitAll(WsMessage.JOB_WAITING, {
+            processId,
+            queuePosition,
+            jobId,
+            orchestratorProcessInstanceId,
         });
+        await this.queueMessageService.sendQueueMessage(QueueEventType.UPDATE, job, queuePosition);
     }
 
     @OnQueueCompleted()
     onComplete(job: Job) {
+        this.logger.log(`[Q Completed] Job ${job.id} is completed`);
+
+        this.queueService.clearQueueTimer(job.id);
+
         this.uiGateway.server.emit(WsMessage.REMOVE_WAITING_SCHEDULE, job);
+
+        const processId = job?.data?.process?.id;
+        this.uiGateway.emitAll(WsMessage.PROCESS_START_COMPLETED, { processId });
     }
 
     @OnQueueFailed()
-    onFailed(job: Job, err: Error) {
-        this.uiGateway.server.emit(WsMessage.REMOVE_WAITING_SCHEDULE, job);
+    async onFailed(job: Job, error: Error) {
         this.logger.error(
             `[Q Fail] Job "${job.data.process.name}" failed`,
-            err
+            error
         );
+
+        this.queueService.clearQueueTimer(job.id);
+
+        this.uiGateway.server.emit(WsMessage.REMOVE_WAITING_SCHEDULE, job);
+
+        const processId = job?.data?.process?.id;
+        const errorMessage = error.message;
+        this.uiGateway.emitAll( WsMessage.JOB_FAILED, { processId, errorMessage });
+        await this.queueMessageService.sendQueueMessage(QueueEventType.FAILED, job);
+    }
+
+    @OnQueueRemoved()
+    async onRemove(job: Job) {
+        this.logger.log(`[Q Removed] Job ${job.id} is removed from queue`);
+
+        this.queueService.clearQueueTimer(job.id);
+
+        this.uiGateway.server.emit(WsMessage.REMOVE_WAITING_SCHEDULE, job);
+
+        await this.queueMessageService.sendQueueMessage(QueueEventType.REMOVE, job);
+
+        const active = await this.queueService.getActiveJobs();
+        const waiting = await this.queueService.getWaitingJobs();
+        const jobs = [...active, ...waiting];
+
+        await Promise.all(jobs.map((job, jobIndex) => {
+            const queuePosition = jobIndex + 1;
+            const processId = job?.data?.process?.id;
+            const clientId = job?.data?.clientId;
+            const jobId = job?.id;
+            if (queuePosition === 1) return;
+
+            this.uiGateway.emitClient(clientId, WsMessage.JOB_WAITING, { processId, queuePosition, jobId });
```

Comment:
WsMessage.JOB_WAITING also for active jobs?

Reply:
I omitted a job that is active